src/tpm.rs

```diff
@@ -83,26 +87,85 @@ pub(crate) fn create_ek(
     let handle = ek::create_ek_object(context, alg, DefaultKey)?;
     let cert = ek::retrieve_ek_pubcert(context, alg)?;
     let (tpm_pub, _, _) = context.read_public(handle)?;
+    let tpm_pub_vec = pub_to_vec(tpm_pub);
 
-    // Convert TPM pub object to Vec<u8>
-    // See: https://github.com/fedora-iot/clevis-pin-tpm2/blob/master/src/tpm_objects.rs#L64
-    let mut offset = 0u64;
-    let mut tpm_pub_vec = Vec::with_capacity((tpm_pub.size + 4) as usize);
-
-    unsafe {
-        let res = Tss2_MU_TPM2B_PUBLIC_Marshal(
-            &tpm_pub,
-            tpm_pub_vec.as_mut_ptr(),
-            tpm_pub_vec.capacity() as u64,
-            &mut offset,
-        );
-        if res != 0 {
-            panic!("out of memory or invalid data received from TPM"); //#[allow_ci]
+    Ok((handle, cert, tpm_pub_vec))
+}
+
+// Multiple TPM objects need to be marshaled for Quoting. This macro will
+// create the appropriate functions when called below. The macro is intended
+// to help with any future similar marshaling functions.
+macro_rules! create_marshal_fn {
+    ($func:ident, $tpmobj:ty, $marshal:ident) => {
+        fn $func(t: $tpmobj) -> Vec<u8> {
+            let mut offset = 0u64;
+            let size = std::mem::size_of::<$tpmobj>();
+
+            let mut tpm_vec = Vec::with_capacity(size);
+            unsafe {
+                let res = $marshal(
+                    &t,
+                    tpm_vec.as_mut_ptr(),
+                    tpm_vec.capacity() as u64,
+                    &mut offset,
+                );
+                if res != 0 {
+                    panic!("out of memory or invalid data from TPM"); //#[allow_ci]
+                }
+
+                // offset is a buffer, so after marshaling function is called it holds the
+                // number of bytes written to the vector
+                tpm_vec.set_len(offset as usize);
+            }
+
+            tpm_vec
         }
-        tpm_pub_vec.set_len(offset as usize);
-    }
+    };
+}
 
-    Ok((handle, cert, tpm_pub_vec))
+// These marshaling functions use the macro above and are based on this format:
+// https://github.com/fedora-iot/clevis-pin-tpm2/blob/main/src/tpm_objects.rs#L64
+// ... and on these marshaling functions:
+// https://github.com/parallaxsecond/rust-tss-esapi/blob/main/tss-esapi-sys/src/ \
+// bindings/x86_64-unknown-linux-gnu.rs#L16010
+//
+// Functions can be created using the following form:
+// create_marshal_fn!(name_of_function_to_create, struct_to_be_marshaled, marshaling_function);
+//
+create_marshal_fn!(pub_to_vec, TPM2B_PUBLIC, Tss2_MU_TPM2B_PUBLIC_Marshal);
```

Comment:
I confuse myself with macros. Is `create_marshal_fn` creating `pub_to_vec` which is used later?

Reply:
@ashcrow Yeah, exactly - there are slightly different functions that all have the same basic structure. In all of the functions, some struct `TPM2B_something` needs to be turned into a vector and has a corresponding marshaling function. The macro can create all of these functions without replicating a bunch of code with only slight variations.

`pub_to_vec` and the other functions created here are used in the next PR on top of this one.